File: roman_int.py
# Roman to Integer
# Given a roman numeral, convert it to an integer.
# Input is guaranteed to be within the range from 1 to 3999.

class Solution:
    # @return an integer
    def romanToInt(self, s):
        dic = {"M":1000, "D":500, "C":100, "L":50, "X":10, "V":5, "I":1}
        res = 0
        prev = None
        
        for key in s:
            if prev!=None and prev<dic[key]:
                res -= 2*prev
            res += dic[key]
            prev = dic[key]
        
        return res

File: roman_int.cpp
class Solution {
    
private:
    int val[255];
    void init() {
        val['I'] = 1; val['V'] = 5; val['X'] = 10; 
        val['L'] = 50; val['C'] = 100; val['D'] = 500; 
        val['M'] = 1000; 
    }
    
public:
    int romanToInt(string s) {
        init();
        int res = 0;
        for (int i=0; i<s.size();i++) {
            if (i>0 && val[s[i-1]]<val[s[i]]) {
                res -= 2*val[s[i-1]];
            }
            res += val[s[i]];
        }
    return res;
    }
};
